### PR TITLE
[FLINK-8739] [table] Optimize DISTINCT aggregates to use the same distinct accumulator if possible

### PR DESCRIPTION
**Pull request by dianfu**
## What is the purpose of the change

*This pull request optimizes the DISTINCT aggregates such as `SELECT COUNT(DISTINCT(a)), SUM(DISTINCT(a))` to use the same distinct accumulator for performance optimization.*

## Brief change log

  - *Separate DistinctAccumulator and the actual accumulator*
  - *Optimize the AggregateCodeGeneration to reuse the same DistinctAccumulator if possible*

## Verifying this change

  - *Existing tests, such as SqlITCase#testDistinctAggOnRowTimeTumbleWindow, AggregateITCase, OverWindowITCase*
  - *Updated harness test GroupAggregateHarnessTest#testDistinctAggregateWithRetract*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

**Comment by dianfu** [comment]
@walterddr Thanks a lot for your comments, very appreciated! Yes, you are right that this PR depends on #7253 for harness test. Otherwise it is impossible to verify whether the distinct accumulator is really reused as the GeneratedAggregate for harness test is not the one code generated by AggregateCodeGenerator.

**Comment by dianfu** [comment]
Thanks a lot for the remind. Have updated the PR.  @walterddr

**Comment by sunjincheng121** [comment]
@dianfu Thanks for the PR. I'll take look at the change after you rebase the code.  :)

**Comment by dianfu** [comment]
@sunjincheng121  Thanks a lot for reviewing this PR. I have rebased the PR.

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/table/runtime/harness/HarnessTestBase.scala [hunk 1]
```diff
@@ -553,6 +384,20 @@ class HarnessTestBase extends StreamingWithStateTestBase {
...
   }
 
+  def getGeneratedAggregationFields(
+      operator: AbstractUdfStreamOperator[_, _],
+      funcName: String,
+      funcClass: Class[_]): Array[Field] = {
+    val function = funcClass.getDeclaredField(funcName)
+    function.setAccessible(true)
+    val generatedAggregation =
+      function.get(operator.getUserFunction).asInstanceOf[GeneratedAggregations]
```
Comment (sunjincheng121) [inline flink-libraries/flink-table/src/test/scala/org/apache/flink/table/runtime/harness/HarnessTestBase.scala]: Can we add a test case that check is the member variable of the test `GeneratedAggregations meeting the expectations? i.e: check the number of test variables or `NonWindowedAggregationHelper` which named `acc[Num]_distinctValueMap_dataview` ?

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/AggregateUtil.scala [hunk 2]
```diff
@@ -1308,6 +1335,7 @@ object AggregateUtil {
     *
     * @param aggregateCalls calcite's aggregate function
     * @param aggregateInputType input type of given aggregates
+    * @param inputFieldsCount number of input fields,
```
Comment (sunjincheng121) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/AggregateUtil.scala]: remove the `,`.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/MatchCodeGenerator.scala [hunk 3]
```diff
@@ -651,7 +651,7 @@ class MatchCodeGenerator(
 
   class AggBuilder(variable: String) {
 
-    private val aggregates = new mutable.ListBuffer[RexCall]()
+    private val  aggregates = new mutable.ListBuffer[RexCall]()
```
Comment (sunjincheng121) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/MatchCodeGenerator.scala]: code format: remove extra spaces.

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/table/runtime/stream/table/AggregateITCase.scala [hunk 4]
```diff
@@ -119,6 +119,39 @@ class AggregateITCase extends StreamingWithStateTestBase {
     assertEquals(expected.sorted, StreamITCase.retractedResults.sorted)
   }
 
+  @Test
+  def testDistinctAGGWithDifferentArgumentOrder(): Unit = {
```
[the remark this answers is not in the document]
Comment (dianfu) [reply flink-libraries/flink-table/src/test/scala/org/apache/flink/table/runtime/stream/table/AggregateITCase.scala]: Make sense.

**Comment by sunjincheng121** [comment]
Hi @dianfu Thanks for the quickly update! 
LGTM. +1 to merged.

**Comment by sunjincheng121** [comment]
Merging...

**Pull request closed without merging**